### PR TITLE
Chunk taskArns into groups of 100

### PR DESCRIPTION
**Pull request by owen**
If the ECS cluster has > 100 tasks, passing them to
ecs.DescribeTasksRequest() will result in the AWS API returning
errors.

This patch breaks them into chunks of at most 100, and calls
DescribeTasks for each chunk.

### CODE REVIEW

File: provider/ecs.go [hunk 1]
```diff
@@ -206,13 +206,30 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
...
+	var splitTaskReq [][]*string
+	for i := 0; i < len(taskArns); i += 100 {
+		sliceEnd := -1
+		if i+100 < len(taskArns) {
+			sliceEnd = i + 100
+		} else {
+			sliceEnd = len(taskArns)
+		}
+		split_taskreq = append(splitTaskReq, taskArns[i:sliceEnd])
+	}
```
Comment (timoreimann) [inline provider/ecs.go]: Can we move the splitting logic into a small function, e.g.,

```go
func chunkedTaskArns(taskArns []*string) [][]*string
```

?

Comment (timoreimann) [inline provider/ecs.go]: Some tests on the newly created function would also be nice to cover the representative cases (e.g., 0, 1, 99, 100, 101, 200, 400).

Comment (owen) [inline provider/ecs.go]: good thinking!  I'll catch that :)

Comment (owen) [inline provider/ecs.go]: also, my git-fu is weak and I managed to commit a broken patch :(  fixes soon.

**Comment by emilevauge** [comment]
@owen shouldn't we rebase this one `v1.2` branch to include it in the next release?

**Comment by timoreimann** [comment]
@owen how's your git-fu going today? ;)

also, do you want to rebase against 1.2 as @emilevauge suggested? The bugfix seems fairly important for ECS users.

**Comment by owen** [comment]
it's _woeful!_ I'm king of the force push I guess.

I've: 
 * rebased against v1.2
 * split to a small helper function
 * added test cases around splittable values

and will be force pushing again as soon as I ensure it passes a `make validate`

Should I keep a PR against master with this code, or do you all typically pull v1.2 fixes back in?  not sure how that typically works for this project.
(and oh look, more git questions, hehe)

**Comment by timoreimann** [comment]
@owen we typically sync back the v1.2 branch with master regularly (usually on release of a release candidate / final version), so there should be no need to maintain a separate branch against master.

File: provider/ecs.go [hunk 2]
```diff
@@ -206,13 +206,30 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
 		taskArns = append(taskArns, req.Data.(*ecs.ListTasksOutput).TaskArns...)
 	}
 
-	req, taskResp := client.ecs.DescribeTasksRequest(&ecs.DescribeTasksInput{
-		Tasks:   taskArns,
-		Cluster: &provider.Cluster,
-	})
+	var splitTaskReq [][]*string
```
Comment (piteur) [inline provider/ecs.go]: Simple comment - as an outsider interrested on this fix:

Is adding a comment to explain why we split the content & do multiple calls will add any value for future maintenance ?
Without knowing the API limitation - or not having the PR context in mind - the `100` split value can be interpreted as coming from nowhere / seems a bit magical.

Comment (timoreimann) [inline provider/ecs.go]: +1 on that one; plus, moving the logic into a dedicated function with an informative name ([as suggested previously](https://github.com/containous/traefik/pull/1209#discussion_r103832525) :-) )

**Comment by owen** [comment]
A test case failure I observed last night made me realize that pointing Traefik to an empty cluster would result in bad AWS API calls being made.  I've gone ahead and picked up that fix by adding an early return after the first call to ListTasks.

This now works for my empty cluster, and my ~hundred instance cluster.  Tests pass, `make validate` looks good, and it only took _one more force push_ to get it to go up 😬 

@timoreimann

**Comment by owen** [comment]
@timoreimann / @emilevauge - I'm wondering if I broke your CI environment with my, er, aggressive Git usage.

Let me know if I can help :(

**Comment by emilevauge** [comment]
@owen yeah, it seems Travis needs a holiday... Could you push force again?

**Comment by owen** [comment]
Looks like the CI is building now.  I added a few more tests to wake up the CI pipeline.

Apologies... I actually just learned that "rebase master -> v1.2" is as easy as a click at the top of the pull request.  That would've skipped at least one force push on my end, heh.
(I have much to learn about GitHub ;-))

File: provider/ecs_test.go [hunk 3]
```diff
@@ -308,3 +308,42 @@ func TestFilterInstance(t *testing.T) {
...
+
+		out := provider.chunkedTaskArns(tasks)
+		var outCount []int
+
+		for _, el := range out {
+			outCount = append(outCount, len(el))
+		}
+
+		if !reflect.DeepEqual(outCount, c.expectedLengths) {
+			t.Fatalf("Chunking %d elements, expected %#v, got %#v (case %d)", c.count, c.expectedLengths, outCount, i)
```
Comment (timoreimann) [inline provider/ecs_test.go]: I think we should do `t.Errorf` here: If one case fails, we still want to see the result of the other cases.

Comment (timoreimann) [inline provider/ecs_test.go]: I also think we can drop the index variable `i` here. `c.count` already suffices to identify the failing case unambiguously.

File: provider/ecs.go [hunk 4]
```diff
@@ -206,13 +206,24 @@ func (provider *ECS) listInstances(ctx context.Context, client *awsClient) ([]ec
 		taskArns = append(taskArns, req.Data.(*ecs.ListTasksOutput).TaskArns...)
 	}
 
-	req, taskResp := client.ecs.DescribeTasksRequest(&ecs.DescribeTasksInput{
-		Tasks:   taskArns,
-		Cluster: &provider.Cluster,
-	})
+	if len(taskArns) == 0 {
```
Comment (owen) [inline provider/ecs.go]: 👍

File: provider/ecs_test.go (hunk 3 above)
Comment (owen) [inline provider/ecs_test.go]: Not sure why I `Fatalf`'ed there rather than `Errorf`'ed - fixed :)

**Comment by timoreimann** [comment]
@owen you haven't go fmt'ed your latest commit -- please run `make fmt` to fix that. Apart from that, LGTM! :heart:

Could you also squash your commits? After that and another successful CI run, I think we are good to go/merge. :-)

**Comment by owen** [comment]
[I fear Travis may not be happy with me again](https://travis-ci.org/containous/traefik/jobs/208818157):

```Unable to checkout gopkg.in/fsnotify.v1
[ERROR]	Update failed for gopkg.in/fsnotify.v1: Cannot detect VCS
[ERROR]	Failed to install: Cannot detect VCS
The command '/bin/sh -c cd integration && glide install' returned a non-zero code: 1
make: *** [build] Error 1
```

**Comment by timoreimann** [comment]
Github is flaking. Retriggered the build.

**Comment by timoreimann** [comment]
Grrr, and now the integration tests are flaking. :(

Retrying HARDER...

**Pull request merged**